### ISSUE TITLE
Pass in pre-collapsed setting for rows

### ISSUE THREAD

**Issue by stguitar** [angular-gantt/angular-gantt#660]
Is there any supported way currently to set a collapsed boolean in my source data that is fed to angular gantt so that rows can me pre-collapsed?

By doing this, it would remove the need to fully render a large data set and let the user get better performance while drilling down to see what they want.

I guess there are 2 options here.  You could leverage the events/methods of collapsing a row to handle this while  managing the data/model on my apps side and just keeping it in sync with what the user does on the gantt chart.    *OR*  modify how that the tree plugin relies on the flag for _collapsed to be on the row, not the row model and therefore would support pre-collapsing of rows.

If others think that this would be nice, and that my second suggestion would work, id be happy to help make it happen.

**Comment by 11joselu** [comment]
Just set this configuration into your app 

``` javascript
app.config(['treeConfig', function(treeConfig) {
      treeConfig.defaultCollapsed = true;
    }])
```

**Comment by stguitar** [comment]
that is very helpful - do you know how I would do this after a user has changed a setting in the application though?

**Comment by 11joselu** [comment]
Do you mean after gantt has been rendered or just a DDBB setting?.

After rendered: just raise tree events 
api.tree.collapseAll()
api.tree.expandAll()

Documentarion: https://www.angular-gantt.com/plugins/tree/

**Comment by stguitar** [comment]
Well, that is why I was curious about "pre-collapse".  Before rendering the gantt chart, in my application, the user may have chosen in their settings that their gantt charts should be collapsed or not. This isnt a system wide setting for us.  It is per user. 

I realize that gantt-tree plugin supports a collapseAll() but this is AFTER its been rendered - I have some users with gantt charts with 635 "projects" each with some number of tasks in them.  The performance to load/render this type of gantt is really pretty bad.  We wanted to offer the user an option to pre-collapse their gantts, or specific gantt charts, to avoid the performance problem. 

So in other words, waiting for the chart to render just to then collapse it, doesn't really help me here.

**Comment by 11joselu** [comment]
Oh sorry, I misunderstood the question. Maybe just passing an custom attribute to de group object and insert a validation into this line [#L270](https://github.com/angular-gantt/angular-gantt/blob/master/src/plugins/tree/tree.controller.js#L270) an emit tree event to collapse that row. It just an idea. 

It's very interesting. I will try to find a solution

**Comment by stguitar** [comment]
No problem. 

I was thinking of leveraging the existing (tree managed) ._collapsed model property if possible. In other words, your inbound datasource could be adjusted to have this property as well.

Right now, angular-gantt sets/manages this property while you collapse/expand a row.  

The tricky part is that your data is passed to angular-gantt, and the tree plugin augments that model with its ._collapsed property so that the model in angular-gantt core doesn't or isn't responsible for collapsing rows - as it shouldn't be.  There doesn't seem to be a way currently to influence this process.

**Comment by 11joselu** [comment]
I just try with this solution:

Tree Controller [#L3](https://github.com/angular-gantt/angular-gantt/blob/master/src/plugins/tree/tree.controller.js#L3). **Inject ganttArrays**, then insert this code **into the controller**

``` javascript
$scope.toCollapse = function(row) {
      if (!angular.isDefined(row.model.children)) {
        return false;
      }

      return arrays.isIn($scope.pluginScope.collapsedRows, row.model.id);
   };
```
**isIn Function**
ganttArrays.service: [#L80](https://github.com/angular-gantt/angular-gantt/blob/master/src/core/logic/util/arrays.service.js#L80)

``` javascript
isIn: function isInArray(arr, item) {

        return ~arr.indexOf(item) !== 0;
  },
```

**treeBodyChildren.tmpl.html**

**Comment by 11joselu** [comment]
I just try with this solution:

Tree.js
[#L15](https://github.com/angular-gantt/angular-gantt/blob/master/src/plugins/tree.js#L15)

```javascript
collpasedRows: '=?'
```

**At the LinkFn**
```javascript

if (scope.collapsedRows === undefined) {
   scope.collapsedRows = [];
 }
```
Tree Controller [#L3](https://github.com/angular-gantt/angular-gantt/blob/master/src/plugins/tree/tree.controller.js#L3). **Inject ganttArrays**, then insert this code **into the controller**

``` javascript
$scope.toCollapse = function(row) {
      if (!angular.isDefined(row.model.children)) {
        return false;
      }

      return arrays.isIn($scope.pluginScope.collapsedRows, row.model.id);
   };
```
**isIn Function**
ganttArrays.service: [#L80](https://github.com/angular-gantt/angular-gantt/blob/master/src/core/logic/util/arrays.service.js#L80)

``` javascript
isIn: function isInArray(arr, item) {

        return ~arr.indexOf(item) !== 0;
  },
```

**treeBodyChildren.tmpl.html**
[#L21](https://github.com/angular-gantt/angular-gantt/blob/master/src/plugins/tree/treeBodyChildren.tmpl.html#L21)

```html
<li ng-repeat="row in childrenRows" ui-tree-node data-collapsed="$parent.toCollapse(row)">
```

**index.html**
[#L270](https://github.com/angular-gantt/angular-gantt/blob/master/demo/app/index.html#L270)

``` html
<gantt-tree enabled="options.sideMode === 'Tree' || options.sideMode === 'TreeTable'"
                                                    header-content="options.treeHeaderContent"
                                                    keep-ancestor-on-filter-row="true"
        collpased-rows=[IDS]
>
```

**I did not try performances issues.**

**Comment by stguitar** [comment]
interesting!  I will have to try that out - it may be a while just as an FYI as I was asking this question in preparation for an upcoming attack run on this feature, as well as finding out if there was an official way to do this (rather than having to create a fork and manage that).  I do not mind doing that though if there is a chance for it to be merged back into the official codebase. 

Anyway, ill let you know how it goes when I begin adding this functionality in a few weeks.

**Comment by Toilal** [comment]
I'm sorry because i'm not much available for this project now, but i'll take time to merge if you manage to write a pull request for this feature :)

**Comment by 11joselu** [comment]
I'm  trying to achieve it to create apPull request but in a root nodes I get this error:
> angular.js: Error: [$compile:ctreq] Controller 'gantt', required by directive 'ganttRowLabel', can't be found!

[#root repeat](https://github.com/11joselu/angular-gantt/blob/tree-pre-collpased/src/plugins/tree/treeBody.tmpl.html#L13l)

**How to reproduce it?**

**index.html add to tree plugin** this attribute: **pre-collapsed-rows="precollapsed"**

Demo **controller** add: **$scope.precollapsed = ["22"]**

**Sample.js to Development** node add ID attribute: { name: "Development", **id: "22"**... }

Still not find a correct solution for this issue.
